### PR TITLE
Add gallery example for MixUp and CutMix

### PR DESCRIPTION
**Pull request by NicolasHug**
Closes https://github.com/pytorch/vision/issues/7766

### CODE REVIEW

File: torchvision/transforms/v2/_augment.py [hunk 1]
```diff
@@ -141,7 +141,7 @@ def _transform(
 
 
 class _BaseMixupCutmix(Transform):
-    def __init__(self, *, alpha: float = 1, num_classes: int, labels_getter="default") -> None:
+    def __init__(self, *, alpha: float = 1., num_classes: int, labels_getter="default") -> None:
```
Comment (NicolasHug) [inline torchvision/transforms/v2/_augment.py:144]: That's a bugfix - bug that I introduced myself when I added the default, and didn't catch because we never tested on the default value (hence the slightly updated test above)

File: gallery/plot_cutmix_mixup.py [hunk 2]
```diff
@@ -1,8 +1,147 @@
...
+that can improve classification accuracy.
+
+These transforms are slightly different from the rest of the Torchvision
+transforms, because they expect
+**batches** of samples as input, not individual images. In this example we'll
+explain how to use them: after the ``DataLoader``, or as part of a collation
+function.
 """
+
+# %%
```
Comment (NicolasHug) [inline gallery/plot_cutmix_mixup.py:18]: Using # `%%` instead of the long `#####...` as in the other examples allows for those scripts to be properly interpreted as notebook within vscode / pycharm. It makes writing / debugging those examples a lot easier. I'm tempted to align the other scripts to use this syntax as well?

File: docs/Makefile [hunk 3]
```diff
@@ -6,7 +6,7 @@ ifneq ($(EXAMPLES_PATTERN),)
 endif
 
 # You can set these variables from the command line.
-SPHINXOPTS    = -W -j auto $(EXAMPLES_PATTERN_OPTS)
+SPHINXOPTS    = -j auto $(EXAMPLES_PATTERN_OPTS)
```
Comment (NicolasHug) [inline docs/Makefile:9]: will put back before merging

File: gallery/plot_cutmix_mixup.py (hunk 2 above)
Comment (pmeier) [reply gallery/plot_cutmix_mixup.py:18]: Yes, sure, but please in a follow-up PR.

File: torchvision/transforms/v2/_augment.py (hunk 1 above)
Comment (pmeier) [reply torchvision/transforms/v2/_augment.py:144]: I'm ok with making the default value the right type, but should we maybe also to `self.alpha = float(alpha)` below? I feel like `MixUp(alpha=2)` is a reasonable thing to do. The user doesn't need to know that internally floats are required.

File: torchvision/transforms/v2/_augment.py [hunk 4]
```diff
@@ -204,13 +204,20 @@ def _mixup_label(self, label: torch.Tensor, *, lam: float) -> torch.Tensor:
 
 
 class Mixup(_BaseMixupCutmix):
-    """[BETA] Apply Mixup to the provided batch of images and labels.
+    """[BETA] Apply MixUp to the provided batch of images and labels.
```
Comment (pmeier) [inline torchvision/transforms/v2/_augment.py:207]: Should we than also rename the classes?

File: gallery/plot_cutmix_mixup.py [hunk 5]
```diff
@@ -1,8 +1,147 @@
...
+    print(f"Before CutMix/MixUp: {images.shape = }, {labels.shape = }")
+    images, labels = cutmix_or_mixup(images, labels)
+    print(f"After CutMix/MixUp: {images.shape = }, {labels.shape = }")
+
+    # <rest of the training loop here>
+    break
+# %%
+#
+# Note how the labels were also transformed: we went from a batched label of
+# shape (batch_size,) to a tensor of shape (batch_size, num_classes). The
```
Comment (pmeier) [inline gallery/plot_cutmix_mixup.py:96]:
```suggestion
# shape ``(batch_size,)`` to a tensor of shape ``(batch_size, NUM_CLASSES)``. The
```

File: gallery/plot_cutmix_mixup.py [hunk 6]
```diff
@@ -1,8 +1,147 @@
...
+for images, labels in dataloader:
+    print(f"Before CutMix/MixUp: {images.shape = }, {labels.shape = }")
+    images, labels = cutmix_or_mixup(images, labels)
+    print(f"After CutMix/MixUp: {images.shape = }, {labels.shape = }")
+
+    # <rest of the training loop here>
+    break
+# %%
+#
+# Note how the labels were also transformed: we went from a batched label of
```
Comment (pmeier) [inline gallery/plot_cutmix_mixup.py:95]: Should we also include a comment on why this happens? I mean we don't want to explain the transforms here, but we could mention that they transform the labels from index based to class based to be able to mix.

File: gallery/plot_cutmix_mixup.py [hunk 7]
```diff
@@ -1,8 +1,147 @@
...
+# shape (batch_size,) to a tensor of shape (batch_size, num_classes). The
+# transformed labels can still be passed as-is to a loss function like
+# :func:`torch.nn.functional.cross_entropy`.
+#
+# As part of the collation function
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# Passing the transforms after the DataLoader is the simplest way to use CutMix
+# and MixUp, but one disadvantage is that it does not take advantage of the
+# DataLoader multi-processing. For that, we can pass those transforms as part of
```
Comment (pmeier) [inline gallery/plot_cutmix_mixup.py:105]: I've glossed over that the first time you brought it up. The collation happens on the main thread, right? If so, what advantage do we have by using the function as part of the collation function?

File: gallery/plot_cutmix_mixup.py [hunk 8]
```diff
@@ -1,8 +1,147 @@
...
+    break
+
+# %%
+# Non-standard input format
+# -------------------------
+#
+# So far we've used a typical sample structure where we pass ``(images,
+# labels)`` as inputs. MixUp and CutMix will magically work by default with most
+# common sample structures: tuples where the second parameter is a tensor label,
+# or dict with a "label[s]" key. Look at the documentation of the
```
Comment (pmeier) [inline gallery/plot_cutmix_mixup.py:130]:
```suggestion
# or dict with a ``"label[s]"`` key. Look at the documentation of the
```

File: gallery/plot_cutmix_mixup.py (hunk 7 above)
Comment (NicolasHug) [reply gallery/plot_cutmix_mixup.py:105]: collation happens on the worker loop https://github.com/pytorch/pytorch/blob/1a661639f77a172df5d1ccd6987049292c6f3440/torch/utils/data/_utils/worker.py#L209

File: gallery/plot_cutmix_mixup.py (hunk 8 above)
Comment (NicolasHug) [reply gallery/plot_cutmix_mixup.py:130]: I don't think we need to wrap quotes into double backticks. It's not code.

File: torchvision/transforms/v2/_augment.py (hunk 4 above)
Comment (NicolasHug) [reply torchvision/transforms/v2/_augment.py:207]: thought about it, went for consistency with timm (i.e. `Cutmix`). No strong opinion.

File: gallery/plot_cutmix_mixup.py (hunk 5 above)
Comment (NicolasHug) [reply gallery/plot_cutmix_mixup.py:96]: I don't think we need to use `NUM_CLASSES`. `num_classes` is short for "the number of classes" here, it doesn't refer to the `NUM_CLASSES` variable.

I also don't think we should add backticks because that will render terribly with the pytorch sphinx theme (https://github.com/pytorch/pytorch_sphinx_theme/issues/107)

File: gallery/plot_cutmix_mixup.py (hunk 6 above)
Comment (NicolasHug) [reply gallery/plot_cutmix_mixup.py:95]: I'm not sure what you mean

> I mean we don't want to explain the transforms here

agreed

> but we could mention that they transform the labels from index based to class based to be able to mix

isn't that what this sentence is explaining already?

File: torchvision/transforms/v2/_augment.py (hunk 4 above)
Comment (pmeier) [reply torchvision/transforms/v2/_augment.py:207]: I don't mind either, but I want both our documentation and class name to be the same. Your choice.

File: gallery/plot_cutmix_mixup.py (hunk 6 above)
Comment (pmeier) [reply gallery/plot_cutmix_mixup.py:95]: Let me rephrase: we don't want to explain the transformations in detail here, i.e. what happens exactly in Mixup or Cutmix, but I feel a small sentence about mixing labels only being possible in probability based labels would make this a little more clear why this happens. Right now, a user could think that we are pointing them toward a implementation detail or quirk rather than a core feature of such a transform.

Of course none of this applies if we expect the user to know what Cutmix and Mixup are and how they work theoretically. In that case, you can dismiss my suggestion.

File: torchvision/transforms/v2/_augment.py (hunk 4 above)
Comment (NicolasHug) [reply torchvision/transforms/v2/_augment.py:207]: :smiling_face_with_tear: https://github.com/pytorch/vision/pull/7731#discussion_r1277216830 I'm going to let you guys decide what to do as I want to avoid re-undo stuff.

(I don't think we absolutely have to align docs and code: CutMix is the technic, Cutmix is the class object - it's OK to have a distinction.)

File: gallery/plot_cutmix_mixup.py (hunk 6 above)
Comment (NicolasHug) [reply gallery/plot_cutmix_mixup.py:95]: I'm personally struggling to find a short sentence explaining why this happens. I have a hard time coming up with something short enough that wouldn't just bring a ton of additional questions. I feel like we should either explain everything, or nothing (i.e. expect users to understand what CutMix is). But if you can provide a suggestion, I'll add it.

File: torchvision/transforms/v2/_augment.py (hunk 4 above)
Comment (pmeier) [reply torchvision/transforms/v2/_augment.py:207]: @vfdev-5 Could you clarify if you meant to only switch to `MixUp` / `CutMix` for the documentation or for the classes as well? I prefer to have both the same, but don't care whether it is `Cutmix` or `CutMix`.

File: docs/Makefile (hunk 3 above)
Comment (pmeier) [reply docs/Makefile:9]: Don't forget.

**Review by pmeier — APPROVED** [review]
Thanks Nicolas!

**Comment by NicolasHug** [comment]
Merged, will address whatever we decide on https://github.com/pytorch/vision/pull/7772/#discussion_r1279756962 later
